connectors/sources/gmail.py

```diff
@@ -0,0 +1,352 @@
+        ]:
+            try:
+                await client.ping()
+                self._logger.debug(f"Successfully connected to {service_name}.")
+            except Exception:
+                self._logger.exception(f"Error while connecting to {service_name}.")
+                raise
+
+    def _dls_enabled(self):
+        if self._features is None:
+            return False
+
+        return self._features.document_level_security_enabled()
+
+    def access_control_query(self, access_control):
+        # filter out 'None' values
+        filtered_access_control = list(
+            filter(
+                lambda access_control_entity: access_control_entity is not None,
+                access_control,
+            )
+        )
+
+        return {
+            "query": {
+                "template": {"params": {"access_control": filtered_access_control}},
+                "source": {
+                    "bool": {
+                        "filter": {
+                            "bool": {
+                                "should": [
+                                    {
+                                        "bool": {
+                                            "must_not": {
+                                                "exists": {"field": ACCESS_CONTROL}
+                                            }
+                                        }
+                                    },
+                                    {
+                                        "terms": {
+                                            f"{ACCESS_CONTROL}.enum": filtered_access_control
+                                        }
+                                    },
+                                ]
+                            }
+                        }
+                    }
+                },
+            }
+        }
+
+    def _user_access_control_doc(self, user, access_control):
+        email = user.get(UserFields.EMAIL.value)
+        created_at = user.get(UserFields.CREATION_DATE.value)
+
+        return {
+            "_id": email,
+            "identity": {"email": email},
+            "created_at": created_at or iso_utc(),
+        } | self.access_control_query(access_control)
+
+    def _decorate_with_access_control(self, document, access_control):
+        if self._dls_enabled():
+            document[ACCESS_CONTROL] = list(
+                set(document.get(ACCESS_CONTROL, []) + access_control)
+            )
+
+        return document
+
+    async def get_access_control(self):
+        if not self._dls_enabled():
+            return
+
+        async for user in self._google_directory_client.users():
+            if user:
+                access_control = [user.get(UserFields.EMAIL.value)]
+
+                yield self._user_access_control_doc(user, access_control)
+
+    def _filtering_enabled(self, filtering):
+        return filtering is not None and filtering.has_advanced_rules()
+
+    async def _message_doc_with_access_control(
```

Comment:
The method name is misleading. It includes both message processing and access control.